### PR TITLE
For #24114 - Move all home screen state into the global AppStore

### PR DESCRIPTION
**Pull request by gabrielluong**
Fixe #24114

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/components/appstate/AppState.kt [hunk 1]
```diff
@@ -4,17 +4,53 @@
...
+ * @property showCollectionPlaceholder If true, shows a placeholder when there are no collections.
+ * @property showSetAsDefaultBrowserCard If true, shows the default browser card
+ * @property recentTabs The list of recent [RecentTab] in the [HomeFragment].
+ * @property recentBookmarks The list of recently saved [BookmarkNode]s to show on the [HomeFragment].
+ * @property recentHistory The list of [RecentlyVisitedItem]s.
+ * @property pocketStories The list of currently shown [PocketRecommendedStory]s.
+ * @property pocketStoriesCategories All [PocketRecommendedStory] categories.
+ * Also serves as an in memory cache of all stories mapped by category allowing for quick stories filtering.
  */
 data class AppState(
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/components/appstate/AppState.kt:41]: drive-by: the `AppStore` isn't _only_ for home screen state, so consider if some or all of these should be a nested data structure like a `Home` data class or a better name.

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/components/appstate/AppState.kt:41]: I intentionally made the refactoring changes as little as possible. I think we can revise as necessary after the initial land.

File: app/src/main/java/org/mozilla/fenix/components/appstate/AppAction.kt [hunk 2]
```diff
@@ -16,4 +25,41 @@ sealed class AppAction : Action {
...
+    data class Change(
+        val topSites: List<TopSite>,
+        val mode: Mode,
+        val collections: List<TabCollection>,
+        val showCollectionPlaceholder: Boolean,
+        val recentTabs: List<RecentTab>,
+        val recentBookmarks: List<RecentBookmark>,
+        val recentHistory: List<RecentlyVisitedItem>
+    ) :
+        AppAction()
```
[the remark this answers is not in the document]
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/components/appstate/AppAction.kt:38]: I do plan on revisiting of all the actions as followups and even the states we have.

**Comment by gabrielluong** [comment]
> All looks good to me! I think we have some follow-ups:
> 
> * Create nested state within `AppState`
> * Create nested reducers
> * Decide if we should keep this state up-to-date even if the home fragment is not displayed

Agree with the tasks for the follow-ups. I will file them before we land.

**Comment by gabrielluong** [comment]
> 🚗 I'm wondering if it might make sense to move the `HomeFragmentStore` into `Components` (and maybe remove the `Fragment` part of the name) and make it an additional global store. It seems like it would accomplish the same goal, and I don't think it would add a measurable performance hit to have a second global object. It would allow us to still have distinct differences between Home state and App state instead of a singular location where we start dumping both.

I did bring up your comment with Christian and Petru. I will acknowledge that one theme that kept coming up has always been about grouping or sectioning the states into a 'Home' section. I do believe we will explore some version of this in the future. 

The response I got from Christian:
> I think my preference is still for the AppStore with eventually nested state. Global state being together makes the system easier to understand, there's one less thread involved too / one is reserved for every instance of a store.
> And, then this state can be used in other UI surfaces as well where it could easily become useful.

The main downsize I saw with having a global `HomeStore` was having 2 global stores. The biggest pro about having it in `AppStore` is that it is a bit more robust for other UI surfaces. This is already true for Focus which has Top Sites displayed as a placeholder for the AwesomeBar.

File: app/src/main/java/org/mozilla/fenix/components/appstate/AppAction.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/components/appstate/AppAction.kt:38]: I left it as a follow up. I just wanted a copy and paste refactoring for ease of review.

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlView.kt [hunk 3]
```diff
@@ -50,23 +51,23 @@ internal fun normalModeAdapterItems(
         items.add(AdapterItem.ExperimentDefaultBrowserCard)
     }
 
-    if (topSites.isNotEmpty()) {
+    if (settings.showTopSitesFeature && topSites.isNotEmpty()) {
```
[the remark this answers is not in the document]
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlView.kt:54]: This is because we don't clear the state when we turn off the setting if it was enable prior. This was highlighted in https://github.com/mozilla-mobile/fenix/pull/24138#pullrequestreview-913246170

> Decide if we should keep this state up-to-date even if the home fragment is not displayed
